File: src/util/Reddit.ts
import * as _ from 'lodash';
import * as snoowrap from 'snoowrap';
import { Configuration } from '../config/Configuration';

export class Reddit {
    private static instance: Reddit = new Reddit();
    private config: Configuration = Configuration.getInstance();
    private r: snoowrap;

    private constructor() {
        if (Reddit.instance) {
            throw new Error('Reddit already exists!');
        }
        const pkginfo = require('pkginfo')(module);
        this.r = new snoowrap({
            userAgent: `${this.config.fetch('bot.name.nick')}/${module.exports.version}`,
            clientId: `${this.config.fetch('reddit.client')}`,
            clientSecret: `${this.config.fetch('reddit.secret')}`,
            username: `${this.config.fetch('reddit.username')}`,
            password: `${this.config.fetch('reddit.password')}`,
        });
        Reddit.instance = this;
    }

    // tslint:disable-next-line:function-name
    public static getInstance(): Reddit {
        return Reddit.instance;
    }

    public getRandomSubmission(subreddit: string, callback?: (post: any) => void): Promise<any> {
        return this.r.getRandomSubmission(subreddit).then(post => {
            if (callback) {
                callback(post);
            }
        });
    }
}

File: src/main/kotlin/utility/Reddit.kt
package utility
import config.Configuration
import externals.SnooWrap
import externals.require
import kotlin.js.Promise

class Reddit {

    companion object {
        var shared: Reddit = Reddit()
    }

    private val config = Configuration.shared
    private val thing = JSON.parse<dynamic>("""{
        "userAgent": "${config.get("bot.name.nick")}/v1",
        "clientId": "${config.get("reddit.clientId")}",
        "clientSecret": "${config.get("reddit.clientSecret")}",
        "username": "${config.get("reddit.username")}",
        "password": "${config.get("reddit.password")}"
    }""".trimIndent())
    val snoo = SnooWrap(thing)

    fun getRandomSubmission(subreddit: String, callback: (post: dynamic) -> Unit): Promise<dynamic> {
        return snoo.getRandomSubmission(subreddit).then { post: dynamic ->
            callback(post)
        }
    }
}